xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp

```diff
@@ -1800,33 +2186,16 @@ int CAMLCodec::Decode(uint8_t *pData, size_t iSize, double dts, double pts)
       m_1st_pts = am_private->am_pkt.lastpts;
   }
 
-  // if we have still frames, demux size will be small
-  // and we need to pre-buffer more.
-  double target_timesize = 1.0;
-  if (iSize < 20)
-    target_timesize = 2.0;
-
-  // keep hw buffered demux above 1 second
-  if (GetTimeSize() < target_timesize && m_speed == DVD_PLAYSPEED_NORMAL)
-    return VC_BUFFER;
-
-  // wait until we get a new frame or 25ms,
-  if (m_old_pictcnt == m_cur_pictcnt)
-    m_ready_event.WaitMSec(25);
+  if (!DequeueFrame(m_lastFrame))
+    return VC_ERROR;
 
   // we must return VC_BUFFER or VC_PICTURE,
   // default to VC_BUFFER.
   int rtn = VC_BUFFER;
-  m_player_pts = DVD_NOPTS_VALUE;
-  if (m_old_pictcnt != m_cur_pictcnt)
+  if (m_lastFrame)
   {
-    m_old_pictcnt++;
     rtn = VC_PICTURE;
-    m_player_pts = pts;
-    // we got a new pict, try and keep hw buffered demux above 2 seconds.
-    // this, combined with the above 1 second check, keeps hw buffered demux between 1 and 2 seconds.
-    // we also check to make sure we keep from filling hw buffer.
-    if (GetTimeSize() < 2.0 && GetDataSize() < m_vbufsize/3)
+    if (GetBufferLevel() < 5)
```

Comment:
The other way is to mimic how vda/vtb worked (not present anymore in code). There hw decoder did an aync callback that stuck the decoded frame in a queue. Decode looked at that queue for throttling demux and signaling has a picture.

Could do this in a process loop, provided the codec_poll_cntl still works under V4L. Then the only thing you have to worry about is the long startup latency before decoded frames are available and can can user hw buffer time for that. 

This way, codec always has N frames ready to go when videoplayervideo wants them.